src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -685,29 +689,12 @@ std::shared_ptr<ExpressionResult<Result>> ExpressionEvaluator::_evaluate_cast_ex
     for (auto chunk_offset = ChunkOffset{0}; chunk_offset < static_cast<ChunkOffset>(result_size); ++chunk_offset) {
       const auto& argument_value = argument_result.value(chunk_offset);
 
-      // NOLINTNEXTLINE(bugprone-branch-clone)
       if constexpr (std::is_same_v<Result, NullValue> || std::is_same_v<ArgumentDataType, NullValue>) {
         // "<Something> to Null" cast. Do nothing, this is handled by the `nulls` vector
-      } else if constexpr (std::is_same_v<Result, pmr_string>) {
-        // "<Something> to String" cast. Sould never fail, thus boost::lexical_cast (which throws on error) is fine
-        values[chunk_offset] = boost::lexical_cast<Result>(argument_value);
       } else {
-        if constexpr (std::is_same_v<ArgumentDataType, pmr_string>) {
-          // "String to Numeric" cast
-          // Same as in SQLite, an illegal conversion (e.g. CAST("Hello" AS INT)) yields zero
-          // Does NOT use boost::lexical_cast() as that would throw on error - and we do not do the
-          // exception-as-flow-control thing.
-          if (!boost::conversion::try_lexical_convert(argument_value, values[chunk_offset])) {
-            values[chunk_offset] = 0;
-          }
-        } else {
-          // "Numeric to Numeric" cast. Use static_cast<> as boost::conversion::try_lexical_convert() would fail for
-          // CAST(5.5 AS INT)
-          values[chunk_offset] = static_cast<Result>(argument_value);
-        }
+        values[chunk_offset] = *lossy_variant_cast<Result>({argument_value});
```

Comment:
# NOREVIEW, just highlights for open questions
This shortens the code, and forwards any `boost::bad_lexical_cast` when Strings cannot be casted to the requested data type. Before, this just resulted in a field with value 0. According to the SQL standarsSQL, an exception should be thrown in such cases. I don't know if catching the error and explicitly `Fail()` here would be nicer.

Reply:
Taken from DB Fiddle / Postgres 13:

```SQL
CREATE TABLE t (a TEXT);
INSERT INTO t VALUES('1');
INSERT INTO t VALUES('1.2');

SELECT * FROM t WHERE CAST(a as INT) > 0;
```
--> `Query Error: error: invalid input syntax for type integer: "1.2"`

Reply:
Also something that we might discuss in a larger group.